src/global/global_init.cc

```diff
@@ -515,6 +515,9 @@ int reopen_as_null(CephContext *cct, int fd)
 
 void global_init_postfork_start(CephContext *cct)
 {
+  // reexpand the meta in child process
+  cct->_conf.finalize_reexpand_meta();
```

Comment:
It seems like this is somehow closing the ASOK and never re-opening it. This is causing all the `rbd-mirror` daemon tests to fail.

Reply:
Actually, it's clearing any pre-existing configuration override for the admin socket path and restoring the built-in defaults:

Expected:
```
2020-11-19T08:40:34.070-0500 7fec9e4613c0  5 asok(0x557aaa292000) init /home/jdillaman/tmp.rbd_mirror/cluster2-client.mirror.0.asok
2020-11-19T08:40:34.070-0500 7fec9e4613c0  5 asok(0x557aaa292000) bind_and_listen /home/jdillaman/tmp.rbd_mirror/cluster2-client.mirror.0.asok
```

Actual:
```
2020-11-19T08:40:34.070-0500 7fec9e4613c0  5 asok(0x557aaa292000) init /home/jdillaman/ceph/build/run/cluster2/out/cluster2-client.mirror.0.105633.93985329206352.asok
2020-11-19T08:40:34.070-0500 7fec9e4613c0  5 asok(0x557aaa292000) bind_and_listen /home/jdillaman/ceph/build/run/cluster2/out/cluster2-client.mirror.0.105633.93985329206352.asok
```

Reply:
It seems it didn't expand it before fork() ?

Reply:
AFAICT, it looks like nobody used to call `finalize_reexpand_meta` so it's quite possible that existing method is buggy and resulting in the default config settings overriding the CLI / `ceph.conf` settings.

Reply:
Yeah, checked the code. It's buggy.

The root cause is that, the `common_preinit()` will set the `admin_socket` to `$run_dir/$cluster-$name.$pid.$cctid.asok` and the level is `CONF_DEFAULT`, and at the same time it will save the `"admin_socket : $run_dir/$cluster-$name.$pid.$cctid.asok"` pair in the `may_reexpand_meta` map.

The reexpand will always parse and set the above key pair with the `CONF_OVERRIDE` level:

```
enum {
  CONF_DEFAULT,                                                                                                                           
  CONF_MON,
  CONF_FILE,
  CONF_ENV,
  CONF_CMDLINE,
  CONF_OVERRIDE,
  CONF_FINAL
};
```
I think we should replace the `admin_socket` value from `may_reexpand_meta` with the option in a higher level if the `$pid` existes, or just remove it.

Reply:
Is it possible to record the enum w/ the original value and then inject it back in at the correct level when re-expanding?

Reply:
Yeah, this is what I did locally.

Currently what the `finalize_reexpand_meta()` is doing are:

1, dup the options recored in `may_reexpand_meta ` map to a higher priority level, the `CONF_OVERRIDE`.
2, update the legacy values from the options.
3, run `call_observers()` to call the obervers callback to apply the changes.

For the Step1, we can just inject it back in the same level and then judge whether need to do the refresh and call the observers.

Reply:
Updated it. I just remove injecting the options back to the `_conf->values` with the same level, which make no sense. We need to refresh it any way, because the options saved in the `may_reexpand_meta` map might have already expanded. The refresh will try to update and use the value with the highest priority.